### PR TITLE
Frank!Doc: Restrict attributes, issue 1883

### PR DESCRIPTION
**Pull request by mhdirkse**
This pull request fixes issue #1883. The fix is implemented as follows:
* I create an abstract class AbstractInterfaceRejector.java for finding suppressed attributes or group memberships. The algorithm finds inherited items and omits items inherited from another interface. This logic is put in the model package, not the doclet (wrapper API) package, even though it only works with FrankClass, not FrankElement. There are two concrete subclasses for omitting attributes and omitting group memberships.
* We have a class AncestorChildNavigation to visit cumulative config children or attributes, referencing them by declared groups, cumulative groups or overridden children. I changed the implementation such that rejecting children and omitting overridden children is no longer in separate parts of the algorithm. We only omit overridden children. The child of a parent FrankElement should be rejected by overriding it by a fake child in a derived class. The algorithm then omits the child of the parent class because it is overridden.
* There are a few changes in ElementChild. An abstract boolean method is added to find fake children. The predicates used for filtering children are updated because we now have fake children.
* Method FrankDocModel.createAttributes is updated to use the AbstractInterfaceRejector algorithm to find rejected attributes, and to make fake attributes for the rejected attribute names. This connects the AbstractInterfaceRejector to the child navigation algorithm. FrankDocModel uses the wrapper API to discover the newly created annotations FrankDocIgnoreTypeMembership and NoFrankAttribute. Work is delegated to a helper class AttributeNotRealSetter to keep the createAttributes() method small.
* ElementType and FrankElement are updated to remove FrankElement-s from groups in the JSON. Some logic in FrankDocModel is needed to set the newly added fields. The comments should explain what is happening here.
* I put the FrankDocIgnoreTypeMembership annotation on MessageStoreSender. I have tested the Frank!Doc website by adding a fake attributes on ITransactionalStorage and by adding a fake derived class of MessageStoreSender with a NoFrankAttribute attribute setter. Of course this fake code is not in this pull request.
* A new field is introduced in the JSON for the website. Building the JSON does not happen with AncestorChildNavigation, so the website did not automatically remove rejected attributes. The new field mentions attributes to be ignored. The frontend is updated to show the attributes that are ignored, which means not inherited.

### CODE REVIEW

**Comment by mhdirkse** [comment]
In the header of this pull request I wrote: "ElementType and FrankElement are updated to remove FrankElement-s from groups in the JSON. Some logic in FrankDocModel is needed to set the newly added fields. The comments should explain what is happening here.". This code was more complicated than needed. The field FrankElement.syntax2ExcludedFromTypes does not require a new loop of iterating FrankElement-s. We can just set this field during the construction of the FrankElement-s.

**Comment by mhdirkse** [comment]
Ik heb de frontend getest na deze commits. Het is dus inderdaad nog steeds ready-for-review.

File: frankDoc/src/main/java/nl/nn/adapterframework/frankdoc/model/AttributeExcludedSetter.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+	private Set<String> namesExcludedDueToIgnoredInterface = new HashSet<>();
+
+	AttributeExcludedSetter(FrankClass clazz) {
+		String ignoredInterface = clazz.getJavaDocTag(FrankElement.JAVADOC_IGNORE_TYPE_MEMBERSHIP);
+		if(ignoredInterface != null) {
+			if(StringUtils.isBlank(ignoredInterface)) {
+				log.warn("Javadoc tag {} requires an argument that refers a Java interface", FrankElement.JAVADOC_IGNORE_TYPE_MEMBERSHIP);
+			} else {
+				log.trace("Have Javadoc tag {} that refers to interface [{}]", () -> FrankElement.JAVADOC_IGNORE_TYPE_MEMBERSHIP, () -> ignoredInterface);
```
Comment (nielsm5) [inline frankDoc/src/main/java/nl/nn/adapterframework/frankdoc/model/AttributeExcludedSetter.java:44]: Ik denk dat je 'has' bedoeld, er vanuitgaande dat de mdc de context bied welk element de tag heeft.

File: frankDoc/src/main/java/nl/nn/adapterframework/frankdoc/model/AttributeExcludedSetter.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+				AttributesFromInterfaceRejector rejector = new AttributesFromInterfaceRejector(new HashSet<>(Arrays.asList(ignoredInterface)));
```
Comment (nielsm5) [inline frankDoc/src/main/java/nl/nn/adapterframework/frankdoc/model/AttributeExcludedSetter.java:45]: Misschien dat het mooier is om een extra constructor toe te voegen voor wanneer je maar 1 waarde hebt. Ik zie overigens niet waar de AttributesFromInterfaceRejector met meerdere waardes gebruikt wordt?

File: webapp/src/main/webapp/iaf/frankdoc/views/element-attributes.html [hunk 3]
```diff
@@ -1,5 +1,6 @@
 <div>
 	<h4>Attributes</h4>
+	<p><span ng-if="element.nonInheritedAttributes" suppress-attribute-inheritance></span></p>
```
Comment (nielsm5) [inline webapp/src/main/webapp/iaf/frankdoc/views/element-attributes.html:3]: Je kan deze if beter op het p element zetten. Anders kan je een extra witregel krijgen wanneer niets suppressed wordt.

**Review by nielsm5 — CHANGES_REQUESTED** [review]

File: frankDoc/src/main/java/nl/nn/adapterframework/frankdoc/model/AbstractInterfaceRejector.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+	private static class Inheriteds {
+		Set<String> rejects;
+		Set<String> retains;
+	}
+
+	AbstractInterfaceRejector(Set<String> rejectedInterfaces) {
+		this.rejectedInterfaces = rejectedInterfaces;
+	}
+
+	abstract Set<String> getAllItems(FrankClass clazz);
```
[the remark this answers is not in the document]
Comment (mhdirkse) [reply frankDoc/src/main/java/nl/nn/adapterframework/frankdoc/model/AbstractInterfaceRejector.java:39]: Dit is AbstractInterfaceRejector. Een "item" kan de naam zijn van een attribute of het kan de naam zijn van een FrankDocGroup. Het gaat hier nooit om methods. Deze code ziet attribuut namen, geen method namen. Dus niet "setXyz" maar "xyz".

[the remark this answers is not in the document]
Comment (mhdirkse) [reply frankDoc/src/main/java/nl/nn/adapterframework/frankdoc/model/AbstractInterfaceRejector.java:39]: Ik weet een manier om deze code makkelijker te maken. Class AbstractInterfaceRejector heeft in de PR zoals die nu is een derived class GroupFromInterfaceRejector. Die class berekent van welke types een FrankClass geen member mag zijn. Dat gebeurt nu volgens hetzelfde algoritmes als wordt gebruikt voor attributes. Maar misschien hoeft het niet hetzelfde algoritme te zijn. Je krijgt praktisch hetzelfde effect als je gewoon kijkt of een FrankClass een interface inherit. Dat weet een FrankClass. Als we hiermee tevreden zijn, al gebeurt er in theorie dan misschien net iets anders, dan kan GroupFromInterfaceRejector weg. AbstractInterfaceRejector en AttributesFromInterfaceRejector kunnen worden samengevoegd.

Als we dit doen, dan wordt het mogelijk om over attributes te praten in plaats van items in de samengevoegde class.

File: webapp/src/main/webapp/iaf/frankdoc/views/element-attributes.html (hunk 3 above)
Comment (mhdirkse) [reply webapp/src/main/webapp/iaf/frankdoc/views/element-attributes.html]: Gedaan.

File: frankDoc/src/main/java/nl/nn/adapterframework/frankdoc/model/AttributeExcludedSetter.java (hunk 2 above)
Comment (mhdirkse) [reply frankDoc/src/main/java/nl/nn/adapterframework/frankdoc/model/AttributeExcludedSetter.java]: Gedaan. Er is inderdaad altijd maar één waarde bij het construeren van een AttributesFromInterfaceRejector. Ik heb de constructor die een Set vraagt weggehaald.

File: frankDoc/src/main/java/nl/nn/adapterframework/frankdoc/model/AttributeExcludedSetter.java (hunk 1 above)
Comment (mhdirkse) [reply frankDoc/src/main/java/nl/nn/adapterframework/frankdoc/model/AttributeExcludedSetter.java]: Ik heb de log melding aangepast.

File: frankDoc/src/main/java/nl/nn/adapterframework/frankdoc/model/AncestorChildNavigation.java [hunk 5]
```diff
@@ -30,28 +28,32 @@
...
 	void run(FrankElement start) {
-		this.rejector.init(start);
 		enter(start);
-		overridden = new HashSet<>();
+		overridden = new HashMap<>();
 		addDeclaredGroupOrRepeatChildrenInXsd();
 		while(current.getNextAncestorThatHasChildren(noChildren) != null) {
 			enter(current.getNextAncestorThatHasChildren(noChildren));
-			if(overridden.isEmpty() && rejector.isNoCumulativeRejected(current)) {
+			if(! getOverriddenChilren().stream().anyMatch(childSelector)) {
```
[the remark this answers is not in the document]
Comment (mhdirkse) [reply frankDoc/src/main/java/nl/nn/adapterframework/frankdoc/model/AncestorChildNavigation.java]: Heb ik verbeterd.

File: frankDoc/src/main/java/nl/nn/adapterframework/frankdoc/model/AbstractInterfaceRejector.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (mhdirkse) [reply frankDoc/src/main/java/nl/nn/adapterframework/frankdoc/model/AbstractInterfaceRejector.java:39]: De code zoals die nu in de PR zit heeft een duidelijk voordeel. Je kan type lidmaatschap opnieuw introduceren. Stel dat je een type Interface hebt dat geïmplementeerd wordt door Parent. Parent heeft een derived class Child. Child heeft een ff.ignoreTypeMembership JavaDoc tag die naar Interface verwijst. Je kan Child een derived class GrandChild geven die een "implements Interface" clause heeft. Dan zal GrandChild weer wél bij het type Interface horen.

Van dit corner case heb ik geen unit test.

**Review by nielsm5 — APPROVED** [review]

File: frankDoc/src/main/java/nl/nn/adapterframework/frankdoc/model/AbstractInterfaceRejector.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (mhdirkse) [reply frankDoc/src/main/java/nl/nn/adapterframework/frankdoc/model/AbstractInterfaceRejector.java:39]: Mijn vorige comment klopt niet, omdat FrankElement altijd de excluded types erft van zijn parent. Ik zal een PR maken om GroupFromInterfaceRejector te elimineren. Daarmee kan AbstractInterfaceRejector worden samengevoegd met zijn derived class. Deze class gaat dan alleen nog over attributes en kan daarmee begrijpelijker worden gemaakt.

[the remark this answers is not in the document]
Comment (mhdirkse) [reply frankDoc/src/main/java/nl/nn/adapterframework/frankdoc/model/AbstractInterfaceRejector.java:39]: De PR waar ik aan werk is https://github.com/ibissource/iaf/pull/2146/files.

[the remark this answers is not in the document]
Comment (mhdirkse) [reply frankDoc/src/main/java/nl/nn/adapterframework/frankdoc/model/AbstractInterfaceRejector.java:39]: Ik bedoel #2146
